controller/rats_manager.py

```diff
@@ -0,0 +1,191 @@
+                    AttributeState.RUNNING.value,
+                ],
+            )
+        vocab = get_tokenizer_by_project(project_id).vocab
+        record_set = record.get_missing_rats_records(project_id, 100, attribute_id)
+        chunk = 0
+        while record_set:
+            entries = []
+            for record_item in record_set:
+                docs = get_docs_from_db(project_id, str(record_item.record_id), vocab)
+                attribute_ids = [str(id) for id in record_item.attribute_ids]
+                for col in text_attributes:
+                    if text_attributes[col] in attribute_ids:
+                        entries.append(
+                            RecordAttributeTokenStatistics(
+                                project_id=project_id,
+                                record_id=record_item.record_id,
+                                attribute_id=text_attributes[col],
+                                num_token=len(docs[col]),
+                            )
+                        )
+            if not project.get(project_id):
+                # rats cant be added (e.g. project deleted)
+                break
+            else:
+                general.add_all(entries)
+                general.commit()
+            if chunk % 20 == 0:
+                # ensure session isn't used up to refresh ocasionally
+                session_token = general.remove_and_refresh_session(session_token, True)
+                tokenization_task = tokenization.get(project_id, task_id)
+                __update_progress(
+                    project_id, attribute_id, tokenization_task, initial_count
+                )
+            record_set = record.get_missing_rats_records(project_id, 100, attribute_id)
+            chunk += 1
+        __finalize_statistic_calculation(project_id, user_id, tokenization_task)
+    except Exception:
+        __handle_error(project_id, user_id, tokenization_task)
+    finally:
+        general.remove_and_refresh_session(session_token, False)
+
+
+def __set_up_statistic_calculation(
+    project_id: str, task_id: str, initial_count: int
+) -> RecordTokenizationTask:
+    tokenization_task = tokenization.get(project_id, task_id)
+    tokenization_task.workload = initial_count
+    tokenization_task.state = enums.TokenizerTask.STATE_IN_PROGRESS.value
+    send_websocket_update(
+        project_id, False, ["rats", "state", str(tokenization_task.state)]
+    )
+    general.commit()
+    return tokenization_task
+
+
+def __update_progress(
+    project_id: str,
+    attribute_id: str,
+    tokenization_task: RecordTokenizationTask,
+    initial_count: int,
+) -> None:
+    current_count = record.count_missing_rats_records(project_id, attribute_id)
+    tokenization_task.progress = round(1 - current_count / initial_count, 4)
+    send_websocket_update(
+        project_id, True, ["rats", "progress", str(tokenization_task.progress)]
+    )
+    general.commit()
+
+
+def __finalize_statistic_calculation(
```

Comment:
function is more of a finalize rats calculation since not just statistics are handles (e.g. deletion of duplicate rats)

Reply:
renamed